Source/Initialization/WarpXInitData.cpp

```diff
@@ -231,23 +233,189 @@ WarpX::PostRestart ()
+       Store_parserString("Ez_external_grid_function(x,y,z)",
+                                                    str_Ez_ext_grid_function);
+
+       // Initialize Efield_fp with external function
+       InitializeExternalFieldsOnGridUsingParser(Efield_fp[lev][0].get(),
+                                                 Efield_fp[lev][1].get(),
+                                                 Efield_fp[lev][2].get(),
+                                                 str_Ex_ext_grid_function,
+                                                 str_Ey_ext_grid_function,
+                                                 str_Ez_ext_grid_function,
+                                                 Ex_nodal_flag, Ey_nodal_flag,
+                                                 Ez_nodal_flag, lev);
     }
 
     if (lev > 0) {
         for (int i = 0; i < 3; ++i) {
-            Efield_aux[lev][i]->setVal(E_external_grid[i]);
-            Bfield_aux[lev][i]->setVal(B_external_grid[i]);
-
             current_cp[lev][i]->setVal(0.0);
-            Efield_cp[lev][i]->setVal(E_external_grid[i]);
-            Bfield_cp[lev][i]->setVal(B_external_grid[i]);
+            if (B_ext_grid_s == "constant" || B_ext_grid_s == "default") {
+               Bfield_aux[lev][i]->setVal(B_external_grid[i]);
+               Bfield_cp[lev][i]->setVal(B_external_grid[i]);
+            }
+            if (E_ext_grid_s == "constant" || E_ext_grid_s == "default") {
+               Efield_aux[lev][i]->setVal(E_external_grid[i]);
+               Efield_cp[lev][i]->setVal(E_external_grid[i]);
+            }
+        }
+        if (B_ext_grid_s == "parse_b_ext_grid_function") {
+
+            InitializeExternalFieldsOnGridUsingParser(Bfield_aux[lev][0].get(),
+                                                      Bfield_aux[lev][1].get(),
+                                                      Bfield_aux[lev][2].get(),
+                                                      str_Bx_ext_grid_function,
+                                                      str_By_ext_grid_function,
+                                                      str_Bz_ext_grid_function,
+                                                      Bx_nodal_flag, By_nodal_flag,
+                                                      Bz_nodal_flag, lev);
+
+            InitializeExternalFieldsOnGridUsingParser(Bfield_cp[lev][0].get(),
+                                                      Bfield_cp[lev][1].get(),
+                                                      Bfield_cp[lev][2].get(),
+                                                      str_Bx_ext_grid_function,
+                                                      str_By_ext_grid_function,
+                                                      str_Bz_ext_grid_function,
+                                                      Bx_nodal_flag, By_nodal_flag,
+                                                      Bz_nodal_flag, lev);
```

Comment:
Don't we want to initialize the fine patch?

Reply:
The fine patch is initialized on lines 337 and 357 for the B and E field respectively.